internal/cortex/frontend/transport/handler.go

```diff
@@ -55,20 +58,52 @@ type Handler struct {
 	cfg          HandlerConfig
 	log          log.Logger
 	roundTripper http.RoundTripper
+	lru          *lru.Cache
+	regex        *regexp.Regexp
+	errorExtract *regexp.Regexp
 
 	// Metrics.
 	querySeconds *prometheus.CounterVec
 	querySeries  *prometheus.CounterVec
 	queryBytes   *prometheus.CounterVec
+	totalQueries *prometheus.CounterVec
+	cachedHits   *prometheus.CounterVec
 	activeUsers  *util.ActiveUsersCleanupService
 }
 
+// CacheableError Returns true if response code is in pre-defined cacheable errors list, else returns false
+func CacheableError(statusCode int) bool {
+	for _, errStatusCode := range cacheableResponseCodes {
+		if errStatusCode == statusCode {
+			return true
+		}
+	}
+	return false
+}
+
 // NewHandler creates a new frontend handler.
 func NewHandler(cfg HandlerConfig, roundTripper http.RoundTripper, log log.Logger, reg prometheus.Registerer) http.Handler {
+	var (
+		LRU *lru.Cache
+	)
+
+	LRU = nil
+
+	if cfg.FailedQueryCacheCapacity > 0 {
+		LRU_res, err := lru.New(cfg.FailedQueryCacheCapacity)
+		LRU = LRU_res
```

Comment:
move this line after the `if` below so that `LRU` is `nil` if there is an error while creating a cache.